src/core_plugins/timelion/public/panels/timechart/schema.js

```diff
@@ -355,13 +357,17 @@ export default function timechartFn(Private, config, $rootScope, $compile) {
             legendCaption = $('<caption class="timelionLegendCaption"></caption>');
             legendCaption.html(emptyCaption);
             canvasElem.find('div.legend table').append(legendCaption);
-
             // legend has been re-created. Apply focus on legend element when previously set
             if (focusedSeries || focusedSeries === 0) {
               const $legendLabels = canvasElem.find('div.legend table .legendLabel>span');
               $legendLabels.get(focusedSeries).focus();
             }
           }
+          //chart has been recreated. Append the new chart values
+          chartValues = '<div id="chartValues" style="display: none">' + JSON.stringify(plotConfig) + '</div>';
+          const chartValuesClear = $(emptyChartValues);
```

Comment:
These two lines are actually never used (the same as the `emptyChartValues`)?

Reply:
The first assignment is used to get the latest chart data to append to the div and the second one finds the previous object and replaces it with and empty value in order to clear old values from the DOM. It's done after the pattern from above for the timelionLegendCaption.

Reply:
You are assigning always `emptyChartValue` which is always the empty string from above. Then you reset it's HTML and never do anything with it? Because what you append is always `chartValues` which is this string above. The element assigned to `chartValuesClear` should nver be assigned to the DOM.

Reply:
@timroes, just to be clear .... appending JSON.stringified data to dom is also not safe ? Isn't that always be properly escaped ?

Reply:
@ppisljar no `JSON.stringify` doesn't properly escape and for the example on the review comment you will have something like: `{"xss": "<img onload='...'>"` that, if directly added to the DOM, will create an img tag with the onload function

Reply:
Solution here could be. Just create the element empty and then use `.textContent` to assign the JSON.stringify to it. That way no HTML in there would be evaluated.